### PR TITLE
 group characters stuff

### PR DESCRIPTION
**Pull request by ThomasBreuer**
moved group characters stuff from `Oscar/examples` to `Oscar/src/Groups`, extended it, added tests

### CODE REVIEW

File: src/Groups/group_characters.jl [hunk 1]
```diff
@@ -212,36 +86,134 @@ mutable struct GAPGroupCharacterTable <: GroupCharacterTable
...
+    if tbls == nothing
+      tbls = Dict()
+      AbstractAlgebra.set_special(G, :character_tables => tbls)
     end
 
-    if p != 0
-      isprime(p) || error("p must be 0 or a prime integer")
-      tbl = mod(tbl, p)
+    tbl = get(tbls, p, nothing)
+    if tbl == nothing
```
Comment (fingolfin) [inline src/Groups/group_characters.jl]: Could also use `get!` plus do block (and then of course also modify the end of the `do` block and the final return
```suggestion
    return get!(tbls, p) do
```

File: src/Groups/group_characters.jl [hunk 2]
```diff
@@ -350,11 +339,11 @@ function Base.show(io::IO, tbl::GAPGroupCharacterTable)
     cents_strings = [d[p] for p in primes]
 
     # Compute display format for power maps.
-    names = Vector{String}(GAP.Globals.ClassNames(tbl.GAPTable))
-    pmaps = Vector{Any}(GAP.Globals.ComputedPowerMaps(tbl.GAPTable))
+    names = Vector{String}(GAP.Globals.ClassNames(gaptbl))
+    pmaps = Vector{Any}(GAP.Globals.ComputedPowerMaps(gaptbl))
     power_maps_primes = []
     power_maps_strings = []
```
Comment (fingolfin) [inline src/Groups/group_characters.jl]:
```suggestion
    power_maps_primes = String[]
    power_maps_strings = String[]
```
for type stability

File: src/Groups/group_characters.jl [hunk 3]
```diff
@@ -548,27 +577,33 @@ function Base.:*(chi::GAPGroupClassFunction, psi::GAPGroupClassFunction)
...
 function scalar_product(chi::GAPGroupClassFunction, psi::GAPGroupClassFunction)
     chi.table === psi.table || error("character tables must be identical")
-    return Nemo.fmpz(GAP.gap_to_julia(GAP.Globals.ScalarProduct(chi.values, psi.values)))
+    return Nemo.fmpz(GAP.Globals.ScalarProduct(chi.values, psi.values))
 end
 
-#TODO: chi(g), ...
-
-function Base.:*(n::T, chi::GAPGroupClassFunction) where T <: Integer
+function Base.:*(n::T, chi::GAPGroupClassFunction) where T <: Union{fmpz, Base.Integer}
```
Comment (fingolfin) [inline src/Groups/group_characters.jl]:
```suggestion
function Base.:*(n::IntegerUnion, chi::GAPGroupClassFunction)
```

File: src/Groups/group_characters.jl [hunk 4]
```diff
@@ -548,27 +577,33 @@ function Base.:*(chi::GAPGroupClassFunction, psi::GAPGroupClassFunction)
...
     return GAPGroupClassFunction(chi.table, n * chi.values)
 end
 
-function Base.:^(chi::GAPGroupClassFunction, n::Int)
+function Base.:^(chi::GAPGroupClassFunction, n::T) where T <: Union{fmpz, Base.Integer}
```
Comment (fingolfin) [inline src/Groups/group_characters.jl]:
```suggestion
function Base.:^(chi::GAPGroupClassFunction, n::IntegerUnion)
```

File: test/Groups/group_characters.jl [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  t_a4 = character_table(alternating_group(4))
+  t_a5 = character_table("A5")
+
+  # `print` shows an abbrev. form
+  print(io, t_a4)
+  @test String(take!(io)) == "character_table(Alt( [ 1 .. 4 ] ))"
+
+  # default `show`
+  show(io, t_a4)
+  @test String(take!(io)) == "Alt( [ 1 .. 4 ] )\n\n 2  2  2       .       .\n 3  1  .       1       1\n                        \n   1a 2a      3a      3b\n2P 1a 1a      3b      3a\n3P 1a 2a      1a      1a\n                        \nχ₁  1  1       1       1\nχ₂  1  1 -ζ₃ - 1      ζ₃\nχ₃  1  1      ζ₃ -ζ₃ - 1\nχ₄  3 -1       .       .\n"
```
Comment (fingolfin) [inline test/Groups/group_characters.jl:12]: Alternatively you could test the `show` method via a doctest, that might be more helpful?

**Review by fingolfin — APPROVED** [review]

File: test/Groups/group_characters.jl (hunk 5 above)
Comment (thofma) [reply test/Groups/group_characters.jl:12]: To save the default show output you can just do `sprint(show, "text/plain", t_a4)`.

**Review by fingolfin — APPROVED** [review]

**Pull request merged**